### PR TITLE
ssh: T4716: Ability to configure RekeyLimit data and time

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

Ability to configure SSH `RekeyLimit` data (in Megabytes) and time (in Minutes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4716

## Component(s) name

ssh
## How to test
```
set service ssh rekey data 1024
set service ssh rekey time 60

```
sshd_config
```
vyos@r14# cat /run/sshd/sshd_config | tail -1
RekeyLimit 1024M 60M
[edit]
vyos@r14# 
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: interface-definitions/ssh.xml.in [hunk 1]
```diff
@@ -206,6 +206,39 @@
...
+                  <help>Threshold data in Megabytes</help>
+                  <valueHelp>
+                    <format>u32:1-65535</format>
+                    <description>Megabytes</description>
+                  </valueHelp>
+                  <constraint>
+                    <validator name="numeric" argument="--range 1-65535"/>
+                  </constraint>
+                </properties>
+                <defaultValue>default</defaultValue>
```
Comment (c-po) [inline interface-definitions/ssh.xml.in]: If you set defaultValue to default, thete should be a completionHelp, valueHelp and constraint, too

File: interface-definitions/ssh.xml.in [hunk 2]
```diff
@@ -206,6 +206,39 @@
             </properties>
             <defaultValue>22</defaultValue>
           </leafNode>
```
Comment (c-po) [inline interface-definitions/ssh.xml.in:208]: Please rename this to simply rekey

**Review by c-po — CHANGES_REQUESTED** [review]

File: interface-definitions/ssh.xml.in (hunk 1 above)
Comment (sever-sever) [reply interface-definitions/ssh.xml.in]: Done!

File: interface-definitions/ssh.xml.in (hunk 2 above)
Comment (sever-sever) [reply interface-definitions/ssh.xml.in:208]: Done!

File: interface-definitions/ssh.xml.in (hunk 1 above)
Comment (c-po) [reply interface-definitions/ssh.xml.in:229]: Please also provide a `valueHelp` for `default`

```xml
                  <valueHelp>
                    <format>default</format>
                    <description>foo bar</description>
                  </valueHelp>
```

File: interface-definitions/ssh.xml.in [hunk 3]
```diff
@@ -206,6 +206,47 @@
...
+                    <description>Megabytes</description>
+                  </valueHelp>
+                  <constraint>
+                    <validator name="numeric" argument="--range 1-65535"/>
+                    <regex>(default)</regex>
+                  </constraint>
+                </properties>
+                <defaultValue>default</defaultValue>
+              </leafNode>
+              <leafNode name="time">
```
Comment (c-po) [inline interface-definitions/ssh.xml.in:231]: Please also provide a `valueHelp` for `none`

```xml
<valueHelp>
  <format>none</format>
  <description>foo bar</description>
</valueHelp>
```

**Comment by c-po** [comment]
I wonder if it really makes sense to specify `default | none` for rekey time/volume. IMHO we should stick to `sshd` build in defaults and not expose `default | none` at all via CLI.

**Comment by sever-sever** [comment]
> I wonder if it really makes sense to specify `default | none` for rekey time/volume. IMHO we should stick to `sshd` build in defaults and not expose `default | none` at all via CLI.

Will do

Update, Done!

File: data/templates/ssh/sshd_config.j2 [hunk 4]
```diff
@@ -96,3 +96,7 @@ DenyGroups {{ access_control.deny.group | join(' ') }}
 # sshd(8) will send a message through the encrypted channel to request a response from the client
 ClientAliveInterval {{ client_keepalive_interval }}
 {% endif %}
+
+{% if rekey.data is vyos_defined and rekey.time is vyos_defined %}
```
Comment (c-po) [inline data/templates/ssh/sshd_config.j2]: As both options are required to set this in the ssh configuration file you should probably add a `verify()` step. Else the user might not know why the config item did not appear.

**Review by c-po — CHANGES_REQUESTED** [review]

File: data/templates/ssh/sshd_config.j2 (hunk 4 above)
Comment (dmbaturin) [reply data/templates/ssh/sshd_config.j2]: Is there really a reason to require both options? According to the [manual](https://man.openbsd.org/sshd_config#RekeyLimit), "[RekeyLimit] specifies the maximum amount of data that may be transmitted or received before the session key is renegotiated, **optionally** followed by a maximum amount of time that may pass before the session key is renegotiated.".

**Review by dmbaturin — CHANGES_REQUESTED** [review]

File: interface-definitions/ssh.xml.in [hunk 5]
```diff
@@ -206,6 +206,37 @@
...
             <defaultValue>22</defaultValue>
           </leafNode>
+          <node name="rekey">
+            <properties>
+              <help>SSH session rekey limit</help>
+            </properties>
+            <children>
+              <leafNode name="data">
+                <properties>
+                  <help>Threshold data in Megabytes</help>
```
Comment (dmbaturin) [inline interface-definitions/ssh.xml.in]: We shouldn't capitalize units. Unless Megabyte is someone's name, it should be lowercase. ;)
Same goes for minutes.

**Review by dmbaturin — APPROVED** [review]

File: data/templates/ssh/sshd_config.j2 (hunk 4 above)
Comment (sever-sever) [reply data/templates/ssh/sshd_config.j2]: Added check for data limit and as an option time limit

File: interface-definitions/ssh.xml.in (hunk 5 above)
Comment (sever-sever) [reply interface-definitions/ssh.xml.in]: Fixed

**Review by dmbaturin — APPROVED** [review]

**Review by c-po — APPROVED** [review]

**Pull request merged**